### DISCUSSION TITLE
CSV Imports with Regex for the read_csv `columns` Parameter

### DISCUSSION THREAD

**Discussion by earlev4** [duckdb/duckdb#11909] [Ideas]
Hi! I know there has been excellent progress with implementing regex with `SELECT COLUMNS`. I'm exploring the possibility of using regex with the columns parameter in read_csv to handle data types dynamically.

Current approach:
```sql
SELECT * FROM read_csv('flights.csv',
    delim = '|',
    header = true,
    columns = {
        'FlightDate': 'DATE',
        'UniqueCarrier': 'VARCHAR',
        'OriginCityName': 'VARCHAR',
        'DestCityName': 'VARCHAR'
    });
```

Proposed approach:
```sql
SELECT * FROM read_csv('flights.csv',
    delim = '|',
    header = true,
    columns = {
        'FlightDate': 'DATE',
        'UniqueCarrier': 'VARCHAR',
        '.*Name.*': 'VARCHAR'
    });
```

**Issue:**
To my understanding, the `read_csv` function requires manually specifying data types for each column when inconsistencies are expected across multiple CSV files. Manually defining each column's data type in the CLI can be cumbersome (repeatedly specifying each column and data type), especially with large datasets like the Backblaze dataset (~90 CSVs, one for each day in a quarter), which features ~148 columns prefixed with `smart_` (e.g., `smart_1_normalized, smart_1_raw, ... smart_255_normalized, smart_255_raw`). Although the desired data type for the smart_ columns is BIGINT, it sometimes varies between BIGINT and VARCHAR depending on whether values or nulls are encountered in the samples.

**Challenge:** 
The `read_csv` function determines the data type from the first CSV file read in a glob operation. For instance, even with `sample_size=-1`, the [Q1 2022 dataset](https://f001.backblazeb2.com/file/Backblaze-Hard-Drive-Data/data_Q1_2022.zip) infers `smart_175_normalized` as `VARCHAR`, whereas the [Q4 2023 dataset ](https://f001.backblazeb2.com/file/Backblaze-Hard-Drive-Data/data_Q4_2023.zip) infers it as `BIGINT`. This inconsistency presents challenges in achieving data type uniformity across different datasets.

**Possible Solutions:**

- Regex `columns` Parameter: Implement regex to dynamically match patterns in column names when assigning data types (e.g., `read_csv('/path/*.csv', columns = {'.*Name.*': 'VARCHAR'})`.
- Sampling Multiple Files: Enhance the CSV sniffer to sample across multiple files and specify file count (e.g., `read_csv('/path/*.csv', sample_files=3, sample_size=-1)`. A limitation is that this approach does not guarantee the CSV sniffer will encounter the desired values. This might be helpful in different scenarios.
- Specifying a Sniff File: Allow users to specify a particular file for the sniffer to evaluate (e.g., `read_csv('/path/*.csv', sniff_file='/path/myfile.csv', sample_size=-1)`. A limitation is that this approach requires knowing which file most represents the data types. Again, this might be helpful in different scenarios.

Enabling regex for the `columns` parameter during `read_csv` would provide a more robust solution by ensuring consistent data types without manually specifying each column.

To summarize, some desired options:
- Use regex with the `columns` parameter
- Scan multiple CSV files with the CSV sniffer
- Specify a file for the CSV sniffer to scan

**Thank you very much for your consideration! I am always appreciative and grateful for the excellent work by the DuckDB community.**

**Comment by pdet** [comment]
Hi @earlev4, Is there any other system that implements the regex solution you proposed? At first glance, it sounds pretty specific to the problem you are having.

I think the most generic solution would be adding an extra option, `sample_files=1`, which would sample the bind types on multiple files, as you described.

We already do multiple runs of the sniffer outside the binding phase. A more complete solution would be to do this in parallel on all files while binding, but that's also a bigger change.

**Comment by earlev4** [comment]
Hi @pdet! Great to hear from you, Pedro! I sincerely appreciate your response and insight! 

Currently, I am not aware of any other systems that implement the proposed regex solution. However, I implemented a workaround using PyArrow. By utilizing PyArrow's `dataset`, I lazily scan the CSV files in a directory to extract the schema. Columns that match the smart pattern are identified using regex and mapped to `int64` within a dictionary. This dictionary is then used to set up `ConvertOptions`, which configures `CsvFileFormat`. The configured `CsvFileFormat` is then used to perform another lazy scan of the CSV files, this time applying the defined column types. Finally, the dataset is written to Parquet files. However, I am exploring ways to achieve this in DuckDB (CLI) without Python or dependencies like PyArrow.

I agree; I think the most generic solution is adding an extra option: `sample_files=1`. I see benefits to this approach for other situations, not just the one I am describing. Given the larger sample size of files, I think that approach should capture the correct schema. The multiple runs in parallel on all files while binding sounds very interesting and exciting!

Thank you very much! 

BTW, I thoroughly enjoyed your talk on CSV parsing!

**Reply by pdet** [reply]
Hi, @earlev4. Thanks for the kind words, and I'm happy to hear you enjoyed the talk :-) ! 

I will most likely add this option after the next minor release. After some further consideration, even parallel sniffing wouldn't help much here since it would be focused on the `union_by_name` options.

**Reply by earlev4** [reply]
Hi @pdet! The kind words are well deserved! 😄 

Thank you very much for considering this option and moving forward with implementation. I am very grateful! I can't wait for it to be released.

Thanks again! I appreciate you!

**Reply by earlev4** [reply]
Hi @pdet! 

Congratulations on the DuckDB 1.0.0 release! I wanted to follow up and see if there was any update on this discussion.

Thanks so much!

**Reply by pdet** [reply]
Hi @earlev4. I think this will realistically come to the next release, i.e., v1.1.

I'm currently reworking the sniffer to work in an adaptive fashion, which basically drastically reduces its cost, and it's a necessary step prior to supporting multi-file scanning in the binder.

**Reply by earlev4** [reply]
Thanks so much, @pdet! I sincerely appreciate the update and all your efforts on the CSV reader and sniffer! I am greatly looking forward to seeing the updates! Thanks again!!!

**Reply by earlev4** [reply]
Hi @pdet! 

Congratulations on another release! Several great enhancements! I was just curious if there were still plans to rework the sniffer for multi-file scanning.

Thanks so much!

**Reply by pdet** [reply]
Hi @earlev4, I promise I did not forget about this! 

Unfortunately, no big changes in this topic are coming to `v1.1`.

I recently pushed https://github.com/duckdb/duckdb/pull/13703, which makes the sniffer avoid small files, but I did not finish it up in time for the release.

When that gets merged, I think I can do a little tweak to sniff multiple files at that point, and I might make the number of files and minimum size adjustable options.

**Reply by earlev4** [reply]
Hi @pdet! Thanks so much for the update and all you do!!! I really appreciate it!!!
